### DISCUSSION TITLE
Micropython for the TC32 (a.k.a. those terrible $3 smartwatches from AliExpress)

### DISCUSSION THREAD

**Discussion by davidgiven** [micropython/micropython#15333] [Other]
I asked about this a little while back on a different forum, but the video's done now!

https://www.youtube.com/watch?v=sv58aPvIonw

This thing is based on a Telink TLSR8232 with 512kB of flash and 16kB of RAM, and Micropython runs just fine in it, complete with a screen driver and pin support. You get about 10kB of heap. The hardware supports BLE but I didn't bother trying to make this work because the vendor BLE stack wants to live in RAM, and consumes... 10kB. 

One interesting thing is that the TLSR8232 is not an ARM; it's based on Telink's own TC32 architecture. Which turns out to be ARM Cortex M0 with a couple of bits different in each opcode --- yes, really. Porting Micropython was easy because I could take the existing Thumb code and search-and-replace the instruction names to turn it into TC32 code.

Would there be any interest in getting this upstreamed? AFAICT Telink is fairly popular in the IoT arena, as well as there being a huge number of cheap, nasty smartwatches based on their chips. One big downside is that you need their compiler, which is a port of gcc with no source available. (Although I can't imagine that porting a modern gcc or LLVM would be hard. You just need to modify the assembler to generate TC32 opcodes from Thumb assembly.)

Here's my fork on github: https://github.com/davidgiven/micropython

**Comment by peterhinch** [comment]
I think the chances of the maintainers accepting this are those of a snowball in Hell. Consider the legality: a dodgy ARM port and a hooky compiler with unspecified licensing. Ugh.

This is my opinion based on familiarity with the project - I'm not a maintainer.

**Reply by robert-hh** [reply]
Maybe user supported ports could be added to awesome microphone.

**Reply by scruss** [reply]
do I detected autocorrect and a reference to [Awesome MicroPython](https://awesome-micropython.com/)?

**Comment by rkompass** [comment]
Very impressing video.
I strongly felt the enthusiasm, and at the same time was wondering:
How would the story end if the smartwatch had 32k of RAM and would cost only 20 cents :-)

**Reply by davidgiven** [reply]
Thanks --- feel free to tell several thousand of your closest friends!

32kB of RAM would be a massive difference. In fact, Telink do make TC32 CPUs with up to 64kB of RAM (I found their product matrix: https://products.telink-semi.cn). For anything larger than that they use a RISC-V core instead. Right now it's not possible to, e.g., display an image from a file on flash because all the image libraries I've seen try to decode into into RAM...

Regarding price, if you search AliExpress for LT716, you might see these watches down to about $1. (Including the large-screen 128x128 models. Additional video forthcoming.)

**Reply by davidgiven** [reply]
I have, BTW, just received a $6.50 watch which has a touchscreen, microphone and speaker, fake heart rate sensor, BL 3, 4.2 and 5, powered by a Cortex M0 with 56MB of RAM and 4MB of flash. There's also a link-layer processor which (presumably?) runs the Bluetooth stack, freeing up the main processor? Unfortunately, figuring out how to program it is left as an exercise for the reader...

**Comment by peterhinch** [comment]
The video is well worth watching. That is a most impressive feat of reverse engineering!

**Reply by davidgiven** [reply]
Thank you!

**Reply by glenn20** [reply]
The algorithm fed me the video yesterday, so I got to watch it before I saw this post. Thanks for an informative and entertaining 40 mins.
